src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/authentication/RequireGroupAuthenticator.java

```diff
@@ -0,0 +1,148 @@
+package com.defenseunicorns.uds.keycloak.plugin.authentication;
+
+import java.util.Optional;
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Simple {@link Authenticator} that checks if a user is a member of a given {@link GroupModel Group}.
+ */
+public class RequireGroupAuthenticator implements Authenticator {
+
+    private static final Logger LOGGER = Logger.getLogger(RequireGroupAuthenticator.class.getName());
+
+    @Override
+    public void authenticate(final AuthenticationFlowContext context) {
+        UserModel user = context.getUser();
+        if (user == null) {
+            logAndFail(context, "Invalid user", AuthenticationFlowError.INVALID_USER);
+            return;
+        }
+
+        ClientModel client = context.getAuthenticationSession().getClient();
+        String groupsAttribute = client.getAttribute("uds.core.groups");
+
+        if (groupsAttribute == null || groupsAttribute.isEmpty()) {
+            LOGGER.infof("No groups detected for client %s", client.getName());
+            context.success();
+            return;
+        }
+
+        JsonNode groupsNode = parseGroupsAttribute(groupsAttribute);
+        if (groupsNode == null) {
+            logAndFail(context, "Failed to parse groups JSON", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
+
+        handleGroupAuthentication(context, user, groupsNode);
+    }
```

Comment:
I think I'd consider inlining `handleGroupAuthentication()` (it's only called on one place) and also inlining `logAndFail()`.

This makes it very clear that all code paths call `context.success()` or `context.failure()` before returning

Reply:
updated